mozilla-services/socorro
# Bug 1468810: Add scaffolding for development of React-based report index.

This is the first step towards re-implementing the crash report page using a more modern frontend stack.

This adds a bunch of stuff:

- A view at `/report/index/<crash_id>/new` as the placeholder for the new crash report page.
- A webpack config that builds files in `/webapp-django/staticfiles` and stores the bundles in `/webapp-django/webpack_bundles`.
- [`django-webpack-loader`](https://github.com/owais/django-webpack-loader), which reads the `/webapp-django/webpack-stats.json` file output by webpack to determine which bundles were built. It then makes these bundles available for use within templates via the `render_bundle` function.
- A `webpack` container that runs concurrently with the `processor` and `webapp` in `make run` that re-builds the frontend files when they change. __Significantly, this shares the built JS with the webapp container via the local volume mount, which may not work on Linux__ (I only tested using Docker for Mac).
- React
- A basic JS file that users React to render content to the crash report page.
- A bunch of eslint tweaks to support correct linting of the new JS files.

The diff is huge because the `package-lock.json` changes are huge.

To test this:

- Checkout the changes.
- Run `make build`.
- Run `make run`.
- Visit http://localhost:8000/report/index/044b488b-b0ed-44de-bb9e-d692b0180418/new and observe an empty crash-stats page with a header "New crash report page placeholder".
- Open up `webapp-django/staticfiles/index.js` and modify the text in the `h1` element.
- Refresh the crash report page and check that the header contents were updated

Besides actually doing the reimplementation, there's a few other things left to do for scaffolding:

- Add frontend tests for the webpack bundles
- Figure out why the webpack container doesn't output anything when it rebuilds the JS or hits a build error.
- Inspect how much build time has increased due to the added packages. I suspect `make build` takes longer now due to the extra packages this added, but I haven't directly compared it.

I think this could be mergable (except maybe for the increased build time issue) as it is, since it should be able to be run parallel to normal development.
